text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+            }
+        }
+    }
+}
+
+// `gen fn`
+gen fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {
+    for value in values {
+        if value.is_odd() {
+            yield value * 2;
+        }
+    }
+}
+```
+
+Iterators created with `gen` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).
+See [the unresolved questions][unresolved-questions] for whether `gen` iterators are fused or may behave strangely after having returned `None` once.
+Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New keyword
+
+Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed.
+
+## Returning/finishing an iterator
+
+`gen` blocks must diverge or return the unit type.  Specifically, the trailing expression must be of the unit or `!` type, and any `return` statements in the block must either be given no argument at all or given an argument of the unit or `!` type.
+
+### Diverging iterators
+
+For example, a `gen` block that produces the infinite sequence `0, 1, 0, 1, 0, 1, ...`, will never return `None`
+from `next`, and only drop its captured data when the iterator is dropped:
+
+```rust
+gen {
+    loop {
+        yield 0;
+        yield 1;
+    }
+}
+```
+
+If a `gen` block panics, the behavior is very similar to `return`, except that `next` unwinds instead of returning `None`.
+
+## Error handling
+
+Within `gen` blocks, the `?` operator desugars as follows.  When its
+argument returns a value indicating "do not short circuit"
+(e.g. `Option::Some(..)`, `Result::Ok(..)`, `ControlFlow::Continue(..)`), that
+value becomes the result of the expression as usual.  When its argument
+returns a value indicating that short-circuiting is desired
+(e.g. `Option::None`, `Result::Err(..)`, `ControlFlow::Break(..)`), the value
+is first yielded (after being converted by `From::from` as usual), then the
+block returns immediately.
+
+Even when `?` is used within a `gen` block, the block must return a
+value of type unit or `!`.  That is, it does not return a value of `Some(..)`,
+`Ok(..)`, or `Continue(..)` as other such blocks might.
+
+However, note that when `?` is used within a `gen` block, all `yield`
+statements will need to be given an argument of a compatible type.  For
+example, if `None?` is used in an expression, then all `yield` statements will
+need to be given arguments of type `Option`.
+
+## Fusing
+
+Like `Generators`, `Iterator`s produced by `gen` panic when invoked again after they have returned `None` once.
+This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially
+produce infinite values of the unit type.
```

Comment:
As the RFC no longer aims to be "experimental," this should probably just state the desired eventual behavior?

Reply:
This section was updated.